source/client/factories_impl.cc

```diff
@@ -94,5 +89,43 @@ OutputCollectorPtr OutputCollectorFactoryImpl::create() const {
   }
 }
 
+HeaderSourceFactoryImpl::HeaderSourceFactoryImpl(const Options& options)
+    : OptionBasedFactoryImpl(options) {}
+
+void HeaderSourceFactoryImpl::setRequestHeader(Envoy::Http::HeaderMap& header,
+                                               absl::string_view key,
+                                               absl::string_view value) const {
+  auto lower_case_key = Envoy::Http::LowerCaseString(std::string(key));
+  header.remove(lower_case_key);
+  // TODO(oschaaf): we've performed zero validation on the header key/value.
+  header.addCopy(lower_case_key, std::string(value));
+}
+
+HeaderSourcePtr HeaderSourceFactoryImpl::create() const {
+  // Note: we assume a valid uri.
+  // Also, we can't resolve, but we do not need that.
+  UriImpl uri(options_.uri());
+  Envoy::Http::HeaderMapPtr header = std::make_unique<Envoy::Http::HeaderMapImpl>();
+
+  header->insertMethod().value(envoy::api::v2::core::RequestMethod_Name(options_.requestMethod()));
+  header->insertPath().value(uri.path());
+  header->insertHost().value(uri.hostAndPort());
+  header->insertScheme().value(uri.scheme() == "https"
+                                   ? Envoy::Http::Headers::get().SchemeValues.Https
+                                   : Envoy::Http::Headers::get().SchemeValues.Http);
+  if (options_.requestBodySize()) {
+    header->insertContentLength().value(options_.requestBodySize());
+  }
+
+  auto request_options = options_.toCommandLineOptions()->request_options();
+  if (request_options.request_headers_size() > 0) {
```

Comment:
You can avoid the check for > 0 unless there is going to be something else in the block.

Reply:
Addressed in b51e654